Add Pull Request Template

alkc opened the pull request:
This update adds a [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates#pull-request-templates) to structure testing and quality assurance of the wgs pipeline.

The checklists can hopefully serve as a way of documenting the testing of the pipeline.

The template follows our quality assurance doc "Rutiner för utveckling av bioinformatiska analysflöden och applikationer"

ViktorHy commented:
add singles to wgs as well for major and minor

Jakob37 commented:
OK, one feedback - the Type of change part (with three checkboxes) means that you never will be able to complete the full checklist, which will be unsatisfying. Could that be just a field to fill out instead?

Jakob37 commented:
A thought that came up when thinking a bit more about this. Should we clarify the responsibility of the reviewer? It seems to me that running the code "in someone else's hands" will capture many issues (as seen in your review of my PR for instance). So for larger changes, maybe an external person should not only review the code, but also test-run it?

alkc commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,81 @@
+
+## Major change
+
+- [ ] Stub run completes without errors or new warnings
+- [ ] `onco` run finishes without any new warnings/errors and the results can be loaded into scout
+- [ ] `wgs` trio run finishes without any new warnings/errors  and the results can be loaded into scout
+- [ ] At least one other person has reviewed and approved my code
+- [ ] I have made corresponding changes to the documentation
```
```suggestion
- [ ] Stub run completes without errors or new warnings
- [ ] `onco` run finishes without any new warnings/errors and the results can be loaded into scout
- [ ] `wgs` single run finishes without any new warnings/errors and the results can be loaded into scout
- [ ] `wgs` trio run finishes without any new warnings/errors and the results can be loaded into scout
- [ ] At least one other person has reviewed and approved my code
- [ ] I have made corresponding changes to the documentation
```

alkc commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,81 @@
+
+- [ ] Stub run completes without errors or new warnings
+- [ ] `onco` run finishes without any new warnings/errors and the results can be loaded into scout
+- [ ] `wgs` trio run finishes without any new warnings/errors  and the results can be loaded into scout
```
```suggestion
- [ ] `wgs` single run finishes without any new warnings/errors and the results can be loaded into scout
- [ ] `wgs` trio run finishes without any new warnings/errors and the results can be loaded into scout
```

alkc commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,81 @@
+
+- [ ] Stub run completes without errors or new warnings
+- [ ] `onco` run finishes without any new warnings/errors and the results can be loaded into scout
+- [ ] `wgs` trio run finishes without any new warnings/errors  and the results can be loaded into scout
```
```suggestion
- [ ] `wgs` single run finishes without any new warnings/errors and the results can be loaded into scout
- [ ] `wgs` trio run finishes without any new warnings/errors and the results can be loaded into scout
```

alkc commented:
> A thought that came up when thinking a bit more about this. Should we clarify the responsibility of the reviewer? It seems to me that running the code "in someone else's hands" will capture many issues (as seen in your review of my PR for instance). So for larger changes, maybe an external person should not only review the code, but also test-run it?

That's a good idea. I like the way @mhkc handles it in the [bonsai repo](https://github.com/Clinical-genomics-Lund/bonsai)  (a good example can be found here: https://github.com/Clinical-Genomics-Lund/bonsai/pull/98) with detailed instructions of how to setup a test instance and a suggested testing procedure that is specific to the PR.

The corresponding PR template looks like this:

```{markdown}
A short description of what this PR fixes

### How to setup and perform the tests

1. step one
2. step two

### Expected outcome 

A short description of the expected outcome of the test above.

### Additional context
```

Jakob37 commented:
> That's a good idea. I like the way @mhkc handles it in the [bonsai repo](https://github.com/Clinical-genomics-Lund/bonsai) (a good example can be found here: [Clinical-Genomics-Lund/bonsai#98](https://github.com/Clinical-Genomics-Lund/bonsai/pull/98)) with detailed instructions of how to setup a test instance and a suggested testing procedure that is specific to the PR.

Yes, that looks like a high-quality testing outline!

alkc commented:
I've added a new draft now. Added a section for the PR author to fill in w/ instructions for reviewers. 

I've also added a section where the PR author/maintainer can edit to document in text who has performed the review/tests. Means we can archive the markdown file if needed and maybe even somehow convert/upload to reqtest?

Jakob37 commented:
> I've added a new draft now. Added a section for the PR author to fill in w/ instructions for reviewers.
> 
> I've also added a section (idea lifted from @clinical-genomics/scout) that the PR author/maintainer can edit to document in text who has performed the review/tests. Means we can archive the markdown file if needed, and maybe convert/upload to reqtest?

Sounds nice to try out! I think we will adjust it further when trying out in action, and seeing whether there are parts still missing, or other parts that feel overkill. I am not familiar with reqtest.

Jakob37 commented:
Comment from yesterday's meeting, that could be added to the checklist: Instructions to tag the last branch head (and make sure the master branch prior to merge is tagged with previous commit).

alkc commented:
@Jakob37: Updated the template. Is this what you had in mind?

Jakob37 commented:
> @Jakob37: Updated the template. Is this what you had in mind?

Yes, looks good!

ryanjameskennedy reviewed (approved):
Perhaps a linting check? Otherwise LGTM!

alkc commented:
> Perhaps a linting check? Otherwise LGTM!

Not a bad idea. Though we need to set up an automatic linter first.

I've never used a linter for NF. Do they work well?

Jakob37 commented:
NF-linting sounds like something that would be interesting to look into. I doubt it is worth implementing it at the moment in this pipeline at the moment though, before we either have migrated it to DSL2 or switched to the nf-core pipeline.

alkc commented:
> NF-linting sounds like something that would be interesting to look into. I doubt it is worth implementing it at the moment in this pipeline though, before we either have migrated it to DSL2 or switched to the nf-core pipeline.

Good feedback. It's definitely something we should look at in the future.

ryanjameskennedy commented:
Isn't there just a default python linter? Pylint for example?

Jakob37 commented:
> Isn't there just a default python linter? Pylint for example?

Do you mean that would be used for Nextflow? A Python linter would not work with Groovy I guess. But maybe I misunderstand what you mean.

alkc commented:
> Isn't there just a default python linter? Pylint for example?

Wasn't obvious from your first comment that you meant a lint check for our utility scripts.

It is a good point though. We do have a lot of homebrewed utility scripts (_mostly_ written in perl) living in the `bin` folder of all our pipeline repos. It wouldn't cost much in terms of workload to set up a lint check that is run on any change to a `*.py` or `*.pl` file.

Python linters work pretty well in my experience, but I have no solid opinion about the ones used for perl, such as [perlcritic](https://metacpan.org/dist/Perl-Critic/view/bin/perlcritic).

alkc commented:
I'll try adding linters that run only on updates to .pl and .py files in a separate update. If it works fine then I'll update the template.

Jakob37 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,148 @@
+
+## Patch
+
+- [ ] Stub run completes without errors or new warnings
```
I think the review-bullet should be here as well in some form. At least having someone glance over the code captures many silly errors. Should be little work it is a tiny change.

Jakob37 replied:
Yes, I agree. It makes sense to leave the possibility open. Maybe it could ideally be structured such that it is a "recommended but skippable" step? I guess there will be many non-emergent patches which really should be reviewed.

I think we are on the same page here. I trust your judgement in how to articulate this!

Jakob37 replied:
OK, maybe I am overly cautious here. I am thinking about patches to something like the wgs-pipeline. Doing small changes to a Python-based local script is probably less sensitive, and quick to fix if there are issues. Your approach above of letter the coder decide is perhaps a good balance!

alkc replied:
Here's an example of an unreviewed change made to the pipeline itself: https://github.com/Clinical-Genomics-Lund/nextflow_wgs/pull/77/files

alkc replied:
I believe it's reasonable to keep it optional for changes to Nextflow code or any of the utility scripts as in the linked example above. 

Reworded the comment a bit:

```suggestion
- [ ] Stub run completes without errors or new warnings
<!--
    Remove the comment markers from the section below if you believe that the patch needs to undergo review. 
    (If in doubt, then play it safe and request a review).
-->
<!--
- [ ] At least one other person has reviewed and approved my code
-->
```

Jakob37 replied:
Yes, that is a good example. I totally agree that it is overkill to review that kind of change. Looks good!
